### ISSUE TITLE
taint2 segfaults on taint_host_memcpy when debugging is on

### ISSUE THREAD

**Issue by vigliag** [panda-re/panda#226]
Hi! I've encountered this segfault while trying to debug a problem I was having with taint2 (I'll talk about that issue in details in a separate issue/mail, basically some data which should have been untainted had several labels attached). Can you please have a look at it?

I was running on the i386 version of panda (`qemu-system-i386`), with a windows7SP1 x86 guest, with these arguments to taint2: `taint2:no_tp=true,debug=true`

Also, I don't know if it this is relevant, or if it may affect the symbols the backtrace refers to, but I've rebuilt  the underlying panda executables a couple of time while this analysis was running (it took quite a bit of time). I have however run this test twice, obtaining the same error and backtrace.

This is the backtrace I got from the core file:
```
#0  FastShad::query (addr=<optimized out>, this=0x7fbe94a3f6e8) at /home/vigliag/thesis/panda/panda/plugins/taint2/fast_shad.h:172
#1  taint_host_memcpy (env_ptr=<optimized out>, dest=<optimized out>, src=94199508767888, greg=<optimized out>, gspec=<optimized out>, size=16, labels_per_reg=4)
    at /home/vigliag/thesis/panda/panda/plugins/taint2/taint_ops.cpp:397
#2  0x00007fbe9accb9f6 in ?? ()
#3  0x0000000000000004 in ?? ()
#4  0x0000000000000004 in ?? ()
#5  0x0000003000000000 in ?? ()
#6  0x00007fbedd9ca650 in ?? () at /home/vigliag/thesis/panda/panda/plugins/taint2/fast_shad.h:119 from /home/vigliag/panda/panda/build/i386-softmmu/panda/plugins/panda_taint2.so
#7  0x00007fbedd9cb160 in ?? () at /home/vigliag/thesis/panda/panda/plugins/taint2/taint_ops.cpp:642 from /home/vigliag/panda/panda/build/i386-softmmu/panda/plugins/panda_taint2.so
#8  0x00007fbedd9cb160 in ?? () at /home/vigliag/thesis/panda/panda/plugins/taint2/taint_ops.cpp:642 from /home/vigliag/panda/panda/build/i386-softmmu/panda/plugins/panda_taint2.so
[...]
```

This is what I had in dmesg. Not sure why it is pointing to those .so files though.
```
[5031744.244676] qemu-system-i38[36627]: segfault at 0 ip 00007fc11215475e sp 00007fc0efd26d60 error 4 in libprotobuf-c.so.1[7fc112153000+8000]
[6054908.054206] qemu-system-i38[26836]: segfault at d0 ip 00007f0d273b1a17 sp 00007f0d16a8d550 error 4 in panda_systaint.so[7f0d273a0000+2e000]
```

these are the last lines of output before the crash
```
update_cb: LLVM[50+4] CB 0 -> 0x0, 0 0 -> 0xfffffffffffffff8, 1 0 -> 0
hostcopy: irrelevant
copy: LLVM[60+4] <- LLVM[50] {}; {}; {}; {}; 
update_cb: LLVM[60+4] CB 0 -> 0x0, 0 0xfffffff8 -> 0xfffffff8, 1 0 -> 0
mix: LLVM[70+8] <- 0+8 {}; {}; {}; {}; {}; {}; {}; {}; 
update_cb: LLVM[70+8] CB 0 -> 0x0, 0 0 -> 0, 1 0 -> 0
hostdel: CPUState[16e+1]
pop: 7fbea95ba010
copy: LLVM[13880+8] <- LLVM[0] {}; {}; {}; {}; {}; {}; {}; {}; 
remove: LLVM[13890+4]
push: 7fbea96f2810
mix: LLVM[30+8] <- 0+8 {}; {}; {}; {}; {}; {}; {}; {}; 
update_cb: LLVM[30+8] CB 0 -> 0x0, 0 0 -> 0, 1 0 -> 0
hostcopy: irrelevant
copy: LLVM[50+4] <- LLVM[40] {}; {}; {}; {}; 
update_cb: LLVM[50+4] CB 0 -> 0x0, 0 0 -> 0, 1 0 -> 0
mix: LLVM[60+8] <- 0+8 {}; {}; {}; {}; {}; {}; {}; {}; 
update_cb: LLVM[60+8] CB 0 -> 0x0, 0 0 -> 0, 1 0 -> 0
mcompute: LLVM[70+4] <- 40 + 10 {}; {}; {}; {}; 
copy: LLVM[80+4] <- LLVM[70] {}; {}; {}; {}; 
update_cb: LLVM[80+4] CB 0 -> 0x0, 0 0 -> 0xfffffffffffffff8, 1 0 -> 0
copy: LLVM[90+4] <- LLVM[80] {}; {}; {}; {}; 
update_cb: LLVM[90+4] CB 0 -> 0x0, 0 0xfffffff8 -> 0xfffffff8, 1 0 -> 0
mix: LLVM[a0+8] <- 0+8 {}; {}; {}; {}; {}; {}; {}; {}; 
update_cb: LLVM[a0+8] CB 0 -> 0x0, 0 0 -> 0, 1 0 -> 0
copy: LLVM[b0+8] <- LLVM[60] {}; {}; {}; {}; {}; {}; {}; {}; 
update_cb: LLVM[b0+8] CB 0 -> 0x0, 0 0 -> 0, 1 0 -> 0
copy: LLVM[c0+8] <- LLVM[a0] {}; {}; {}; {}; {}; {}; {}; {}; 
update_cb: LLVM[c0+8] CB 0 -> 0x0, 0 0 -> 0, 1 0 -> 0
hostmemcpy: CPUState[55ac88421480+10] <- CPUState[55ac88421490] (offsets 1d0 <- 1e0) 
```

Unfortunately I had `CPU_LOG_LLVM_IR ` and `CPU_LOG_TB_IN_ASM` off at that moment to limit the output, so I don't have the basic block that was being executed in that moment.

Another interesting thing to note is that that seems to be the first "not unrelevant" hostmemcpy in the execution.

Thanks in advance,
Gabriele

**Comment by moyix** [comment]
Hmm, is there any chance you could share the record/replay log (the -rr-snp and -rr-nondet.log)? Also, for debugging, it may be helpful to rebuild PANDA with `../build.sh --enable-debug`, which will avoid the various `<optimized out>` messages above.

More generally I would guess that the debugging code has simply atrophied a bit and gotten out of sync – @phulin used to when building `taint2` but it hasn't seen a lot of use since then. It would be good to have it working again though!

**Comment by vigliag** [comment]
Okay I may have figured out what's going on (not a problem with the translate_callback)
I can easily reproduce the problem by toggling a comment on `panda_do_flush_tb()` occurring around 2%.

```
without flushing tb at 2%

opening nondet log for read :	./sysrec-rr-nondet.log
translating SYSENTER, pc=1999270066 asid=533528576
translating SYSENTER, pc=1999270066 asid=237596672
./sysrec-rr-nondet.log:   783402885 instrs total.
sysrec:     7834033 (  1.00%) instrs.    3.26 sec.  0.59 GB ram.
sysrec:    15668059 (  2.00%) instrs.    5.54 sec.  0.60 GB ram.
sysrec:    23502088 (  3.00%) instrs.   10.63 sec.  0.63 GB ram.
sysrec:    31336120 (  4.00%) instrs.   14.16 sec.  0.63 GB ram.
sysrec:    39170148 (  5.00%) instrs.   18.85 sec.  0.65 GB ram.


flushing tb at 2%

opening nondet log for read :	./sysrec-rr-nondet.log
translating SYSENTER, pc=1999270066 asid=533528576
translating SYSENTER, pc=1999270066 asid=237596672
./sysrec-rr-nondet.log:   783402885 instrs total.
sysrec:     7834033 (  1.00%) instrs.    3.18 sec.  0.59 GB ram.
sysrec:    15668059 (  2.00%) instrs.    5.56 sec.  0.61 GB ram.
translating SYSENTER, pc=1999270066 asid=460103680
translating SYSENTER, pc=1999270066 asid=460103680
sysrec:    23502088 (  3.00%) instrs.   11.30 sec.  0.61 GB ram.
sysrec:    31336120 (  4.00%) instrs.   15.02 sec.  0.61 GB ram.
sysrec:    39170148 (  5.00%) instrs.   19.93 sec.  0.64 GB ram.
```

From what I see, the problem is about the way syscalls2 (and my plugin) works. I don't know how qemu works exactly, but I would assume it is possible that the same basic block is used by several processes, and translated only once.

If that's the case, then the problem would be that syscalls2 stores, when the instruction is translated, the pc together with the current asid. Then, in the exec_callback(), it doesn't do anything if the asid is not the same of when the block was translated. [relevant code](https://github.com/panda-re/panda/blob/master/panda/plugins/syscalls2/syscalls2.cpp#L420)

I've tried removing the asid check, and it seems to fix the problem (in that I can see all systemcalls without flushing the tb). Or at least, it seems to work with the way systemcalls are made on windows7Sp1 X86. This fix may break things on other OSs though.

**Comment by vigliag** [comment]
Sure I can share the log, although it takes so much time running it with llvm+taint2, and requires so much ram (several hours and several tens of gigabytes of ram), that I was hoping more for someone to give a quick glance to taint_host_memcpy's code, to see if there's something obviously misplaced.

About the broader taint2 issue that I was trying to debug, I've sent you a mail explaining it in details. Sorry it's pretty long, but I wanted to give some context about how I'm using taint2, and ask your opinion about my project in general. I've also tried mailing the same mail to @phulin, but I've only been able to find an old mail address. I can also post it in a separate github issue.

**Comment by moyix** [comment]
Sadly @phulin has moved on to law school, so he's probably not going to have much time to reply. I'll try to take a look and reply to your mail – my general advice with taint system issues is try to come up with a very small repro case and then trace through it. When we've done this maybe 8/10 times it turned out the taint system was correct but produced unexpected results.
